### ISSUE TITLE
File/Folder restructure.

### ISSUE THREAD

**Issue by crowbarmaster** [crowbarmaster/BedrockManagementService#66]
A lot has changed since I put solid thought into this. With the recent change to single file publishing, the directory is much easier to traverse, and the need for a couple of folders is pretty well depreciated. If you fellas want to chime in and give your thoughts on this, your more than welcome to!

**Comment by crowbarmaster** [comment]
* Moving `.\Service\Globals.conf` to `.\Service.conf`. Removing `Service` directory.
* Moving `.\Client\Config.conf` to `.\Client.conf`. Removing `Client` directory.
* Renaming `.\Server` to `.\BmsConfig`.
* Removing all config backup directories. These were more a debug feature, If they were useful, I can bring them back.
* Renaming `.\Server\Configs` to `.\BmsConfig\ServerConfigs`.
* Renaming `.\Server\MCSFiles` to `.\BmsConfig\BDSUpdates`.
* Renaming default server backup path to `.\ServerBackups`.
* Player database folders merged to one `PlayerRecords` folder.

**Comment by tiamat2012** [comment]
Will there be a migration from the old structure to the new one?  Or will I need to manually save a copy and rebuild my server configuration via the client UI?  Perhaps a simple power shell script to move the files around so people don’t have to “start over”?

**Comment by demifiend9** [comment]
@crowbarmaster This would be a good time for changing the naming scheme of backups. Currently you can only know the the time of the backup from backup manager, or through filesystem timestamps which are volatile.

backup_yyyymmdd_hhmmss would be great, sortable and and you can easily glance and know when it was made.

**Comment by tiamat2012** [comment]
Version your backups and export files, so you know exactly what format it was written in to be able to import it later.

**Comment by crowbarmaster** [comment]
> Will there be a migration from the old structure to the new one? Or will I need to manually save a copy and rebuild my server configuration via the client UI? Perhaps a simple power shell script to move the files around so people don’t have to “start over”?
> 
> As more people use this app, you will have to start getting into these sorts of versioning issues. The suggestion about export and importing things like player data will help you, but users won’t want to start over for each version. And then they will tend to leave an old version running forever, missing new features and bug fixes. And the longer they don’t upgrade, the harder the migration. People may have to step up multiple times, or your migration code will be very complex dealing with multiple old-to-new upgrade paths.

Oh yes, absolutely! Not expecting anyone to tackle that on their own. First thing to load is the globals.conf file, if it still exists in its old location, I will programatically move everything. I will leave this code around for a few releases, should be long enough that everyone is up to speed!



> @crowbarmaster This would be a good time for changing the naming scheme of backups. Currently you can only know the the time of the backup from backup manager, or through filesystem timestamps which are volatile.
> 
> backup_yyyymmdd_hhmmss would be great, sortable and and you can easily glance and know when it was made.
> 
> I have a couple other suggestions regarding backups but they'll need a separate issue.

I am aware of the difficulty facing the folder naming scheme for backups. The challenge for me is that `DateTime.Now(string format)` does not have a default format that is path-friendly. Simple enough to date the folders in the way that it appears in log filenames, but parsing the filenames back into dates again was painful. What you see now is the datetime in Ticks. Going along with what @tiamat2012 mentions above, backups will soon be shipped with not only versions, but also this same Tick value, backup size, and any other info I should keep with. With the backups containing such info in a `BackupInfo.json` file, the path can be named easily, since the folders name no longer contains actual information. Export packages will also receive the same treatment.

**Comment by mhensler** [comment]
> I am aware of the difficulty facing the folder naming scheme for backups. The challenge for me is that `DateTime.Now(string format)` does not have a default format that is path-friendly. Simple enough to date the folders in the way that it appears in log filenames, but parsing the filenames back into dates again was painful. What you see now is the datetime in Ticks. Going along with what @tiamat2012 mentions above, backups will soon be shipped with not only versions, but also this same Tick value, backup size, and any other info I should keep with. With the backups containing such info in a `BackupInfo.json` file, the path can be named easily, since the folders name no longer contains actual information. Export packages will also receive the same treatment.

A timestamp is very user-friendly. Ticks don't mean anything to me. Forgive my ignorance, here. I don't know enough to appreciate the challenge in adding the current system time to the backup path.  Are you not able to grab the current time from the system?

**Comment by crowbarmaster** [comment]
Yes I can, and do! However the different arrangements available as a string
are not path friendly, often containing the ":" character, which is illegal
to use in a path. To make haste, I chose to run with Ticks, a long integer
defining the number of ticks from I believe starting from some date in
1970. Datetime provides an easy way to pull the exact date out of this from
a string, and since it's all numbers, it is path friendly. Soon, this info
will be in a file shipped with the backup, meaning the folder can be named
anything, eliminating the need for this worry. I will be free to date the
backup folders in the same manner as the logs then.

**Comment by demifiend9** [comment]
Since we're already talking about backups...

1. How about tagging/differentiating scheduled backups and manual backups? This way the backup limit would only apply to scheduled backups and backups created manually by pressing the button would be exempt from it. You can delete the manual backups from backup manager if you want. 

2. What are your thoughts on zipping the backups? The reason for doing this would be to reduce the large number of small files to one big file. I have my backup folder synced to a cloud service and dealing with large number of files is much slower and inefficient than one big file. Since this isn't about file size you could limit the compression level to very low or even zero to avoid straining the cpu. This is merely a suggestion ofc, absolutely not a necessity.

**Comment by tiamat2012** [comment]
Try

> DateTime.Now.ToString("yyyyMMdd_HHmmss")

or:

$"{DateTime.Now:yyyyMMdd_HHmmss}"

https://docs.microsoft.com/en-us/dotnet/standard/base-types/custom-date-and-time-format-strings

I think this would be better than ticks as well.

I also like the idea of ZIP files, because databases do tend to compress decently.  Early on I thought the backups were Zip'd, and had mentioned the ZipFile class (in error).

https://docs.microsoft.com/en-us/dotnet/api/system.io.compression.zipfile?view=net-6.0

You could pause the server, make the copy of the files with truncation, then release the server for updates (because ZIP'ing could take a little bit of time) like you do today, and finally create the ZIP file and delete the copied files afterward.  Restore would be to unzip the file into the right folder.  BTW - when you do a restore, do you clear the target folder first?  Because there could be new database files for chunks that were not visited in the backed up world being restored, and those should not be present after the world is restored.

**Comment by crowbarmaster** [comment]
> Try
> 
> > DateTime.Now.ToString("yyyyMMdd_HHmmss")
> 
> or:
> 
> > $"{DateTime.Now:yyyyMMdd_HHmmss}"
> 
> https://docs.microsoft.com/en-us/dotnet/standard/base-types/custom-date-and-time-format-strings
> 
> I think this would be better than ticks as well.
> 
> I also like the idea of ZIP files, because databases do tend to compress decently. Early on I thought the backups were Zip'd, and had mentioned the ZipFile class (in error).
> 
> https://docs.microsoft.com/en-us/dotnet/api/system.io.compression.zipfile?view=net-6.0
> 
> You could pause the server, make the copy of the files with truncation, then release the server for updates (because ZIP'ing could take a little bit of time) like you do today, and finally create the ZIP file and delete the copied files afterward. Restore would be to unzip the file into the right folder. BTW - when you do a restore, do you clear the target folder first? Because there could be new database files for chunks that were not visited in the backed up world being restored, and those should not be present after the world is restored.
> 
> You can also use ParseExact to read the date back from the filename:
> 
> https://docs.microsoft.com/en-us/dotnet/api/system.datetime.parseexact?view=net-6.0

It's not a problem these days to adjust the backups to use a custom format. I set the backup system up quite a while ago when this was but a small server launcher, I had much less knowledge of .NET, and limited time. My former green self simply used what I could find built-in to get the job done, instead of taking the time to find something nicer.

I would assume your okay with the custom format MM-dd-yy_hh.mm.ss.ffff used for the logs? This is written the same way as the second example above. It's really only a few short steps to make this happen, and you guys should see this effective in the next update. 

Oh, fear not, ZipFile and ZipArchive are good friends of mind these days 😄 If the goal is to make everything exportable via the client, some zipping will have to happen anyways. Since zipping a folder and sending a file takes longer than simply sending a pre-existing archive, it only makes sense to make this happen. The whole process is actually easier than you detailed - Currently, I read the files detailed in the query to a MemoryStream, which I dump the contents to a byte array, BlockCopy to exact length detailed by the query into a new array to be written to file. Should be easy enough to use ZipArchive here instead! Actually, because the contents are stored in RAM, the action of zipping takes little time more than writing the file.

 You will be happy to hear that I purge the directory before restoring a backup!

**Comment by tiamat2012** [comment]
Year, month, day with leading zeros allows you to sort the file names in the right order so you can do things like delete oldest in windows explorer.  Tradition US dates with Month first will not do an alphabetical sort in chronological order.  So it might be helpful to change up the format.  Eventually you need to think about log retention and setting limits.  Servers tend to run out of disk space because no one ever cleans up old logs.

**Comment by mhensler** [comment]
I'm also a fan of the "yyyyMMdd_HHmmss" format.

I would prune scheduled and manual backups according to a common retention schedule, which you can set at the server instance level (as opposed to the host).

If I fire off a manual backup that I want to hold longer, I'll move the file out of the folder that BMS manages.

**Comment by crowbarmaster** [comment]
> Year, month, day with leading zeros allows you to sort the file names in the right order so you can do things like delete oldest in windows explorer.  Tradition US dates with Month first will not do an alphabetical sort in chronological order.  So it might be helpful to change up the format.  Eventually you need to think about log retention and setting limits.  Servers tend to run out of disk space because no one ever cleans up old logs.

Well you've made a good point with the sorting issue. I will include this change for both backups and logs in the future release. Log retention can be had easily enough!

At one point in time, the max backup count belonged to the server. However, it was the only property attached to the server, and for simplicity sake in getting things working, it went to the service. If you guys would appreciate having the backup system setup as part of the server instead of service wide, it can be so. I also had future plans to one day restrict server updates to a server level, in the event that a server should be "Frozen" to a particular version.

**Comment by tiamat2012** [comment]
There is also the cron schedule, full or world only backups, and destination folder settings to think about.  I don’t have an opinion since I only run 1 server, but I can see people asking for backup settings per server based on what sort of world and activity is running on each one, how much disk space they have, etc.  They may have some that are more read-only parkour and don’t need backups and others that are survivor worlds that need them more frequently.  Maybe you should wait on this until the next big release?

**Comment by crowbarmaster** [comment]
> There is also the cron schedule, full or world only backups, and destination folder settings to think about. I don’t have an opinion since I only run 1 server, but I can see people asking for backup settings per server based on what sort of world and activity is running on each one, how much disk space they have, etc. They may have some that are more read-only parkour and don’t need backups and others that are survivor worlds that need them more frequently. Maybe you should wait on this until the next big release?

Yeah, I would be giving those to the server as well, wouldn't make much sense otherwise. I agree, waiting until the active issues are closed and a new release candidate ships is a must. I think we have a pretty decent list that will make for a nice update to the public. I am going write up what core issues we have, we will get 'em tested and working well, and open a new batch of features to work on! Still, these discussions have been very informative, and many will serve as the basis for aforementioned issues to open!

**Comment by crowbarmaster** [comment]
> * Moving `.\Service\Globals.conf` to `.\Service.conf`. Removing `Service` directory.
> * Moving `.\Client\Config.conf` to `.\Client.conf`. Removing `Client` directory.
> * Renaming `.\Server` to `.\BmsConfig`.
> * Removing all config backup directories. These were more a debug feature, If they were useful, I can bring them back.
> * Renaming `.\Server\Configs` to `.\BmsConfig\ServerConfigs`.
> * Renaming `.\Server\MCSFiles` to `.\BmsConfig\BDSUpdates`.
> * Renaming default server backup path to `.\ServerBackups`.
> * Player database folders merged to one `PlayerRecords` folder.

Additions to list:
* Logs naming changed: `{logType}_MM-dd-yy_hh.mm.ss.ffff` to `(ServerLog\ServiceLog\ClientLog)-yyyyMMdd_HHmmss`.
* Backup naming changed: `Backup_TickCount` to `Backup_yyyyMMdd_HHmmss`

A test will be performed at service or client startup to automatically relocate files, as well as fixing names on existing backups and logs. Test simply checks for Globals.conf to trigger the move. Almost written now, should be debugging shortly!

**Comment by crowbarmaster** [comment]
Currently, The name of the logs are what they are to make sure a new logger runs a new file on startup. Would you guys appreciate it if the logs of past instances were appended to a zip file? Because they are text and usually not too large, they can be added to a zip file with ZipArchive fairly easily and without adding a load of time to startup.

**Comment by crowbarmaster** [comment]
This morning while testing the upgrade assistant, I ran into an issue fixing file names. I have to extend the current time format to `yyyyMMdd_HHmmssff` in lieu that two or more backups or logs where made within the same minute. Failed service launches or the event that someone fired off a series of backup events back-to-back will cause errors and leave old names behind. I can assert that adding 2 digits of milliseconds works well!

**Comment by crowbarmaster** [comment]
Backups are successfully writing to a zip archive today!

![image](https://user-images.githubusercontent.com/69368213/152699114-bbbdd336-95fd-48bc-ab49-0f6840ec6991.png)

Many changes need to be made yet when it comes to rollback, but we are close!

**Comment by crowbarmaster** [comment]
Rollbacks are working! I can't justify doing work on the backend of the backup system without a little frontend work 😉 The backup manager is getting a refresh:

![image](https://user-images.githubusercontent.com/69368213/152712887-c8235504-f262-4843-aade-0d6cfc7bdb2e.png)

I am going to give some thought tonight on menu strip layouts!

**Comment by crowbarmaster** [comment]
Most all of what has been talked about here has been verified working on my end! Probably a few bugs here, so be vigilant! Feel free to open issues on anything that looks out of order! Thanks guys!
